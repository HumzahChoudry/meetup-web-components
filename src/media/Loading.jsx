import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import loaderSprite from '../../assets/svg/loaderSprite.svg';

const SwarmLoader = () => (
	<svg class="swarm-loader" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 180 180">
		<circle id="circle1" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle2" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle3" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle4" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle5" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle6" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle7" cx="90.5" cy="90.5" r="6.5"/>
		<circle id="circle8" cx="90.5" cy="90.5" r="6.5"/>

		<defs>
			<filter id="goo">
				<feGaussianBlur in="SourceGraphic" stdDeviation="10" result="blur" />
				<feColorMatrix in="blur" mode="matrix" values="1 0 0 0 0  0 1 0 0 0  0 0 1 0 0  0 0 0 50 -12" result="goo" />
				<feBlend in="SourceGraphic" in2="goo" />
			</filter>
		</defs>
	</svg>
);

const MaterialLoader = () => (
	<svg class="circular" viewBox="0 0 50 50">
		<circle class="circular-path" cx="25" cy="25" r="20" fill="none" stroke-width="3" stroke-miterlimit="10" />
	</svg>
);

const Loading = ({
	className,
	color,
	size,
	fullCover, // position: fixed // eslint-disable-line no-unused-vars
	partialCover, // position: absolute // eslint-disable-line no-unused-vars
	scrimColor,
	option, // temporary prop
	...other
}) => {
	const classNames = cx(
		'loading',
		{
			'loading--cover': fullCover || partialCover,
			'loading--fullCover': fullCover,
			'loading--partialCover': partialCover
		},
		className
	);
	const renderLoader = (loaderOption) => {
			switch (loaderOption) {
				case 'swarm':
					return (
						[
							<SwarmLoader />
						]
					);
				case 'busySwarm':
					return (
						<div
							className="busySwarm"
							style={{
								mask: `url(${loaderSprite}) no-repeat 0 0`,
								WebkitMask: `url(${loaderSprite}) no-repeat 0 0`,
							}}
						/>
					);
				case 'circular':
					return (
						<MaterialLoader />
					);
				default:
					return (
						<MaterialLoader />
					);

			}
	};

	return(
		<div
			className={classNames}
			style={{
				'--ldr_size': size,
				'--ldr_scrimColor': scrimColor,
				'--ldr_color': color
			}}
		>
			{renderLoader(option)}
		</div>
	);
};

Loading.defaultProps = {
	option: 'swarm',
};

Loading.propTypes = {
	option: PropTypes.string.isRequired,
};

export default Loading;
