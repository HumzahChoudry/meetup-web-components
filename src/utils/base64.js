/**
 * @function getIconAsBase64Uri
 * @param {string} icon name
 * @returns {string} base64-encoded data uri for SVG icon
 */
export const getIconAsBase64Uri = name => {
	try {
		return require(`base64-image!swarm-icons/dist/optimized/${name}.svg`);
	} catch(e) {
		return new Error(`Could not construct base64 uri for icon "${name}" \n ${e}`);
	}
};
