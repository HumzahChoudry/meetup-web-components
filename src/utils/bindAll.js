/**
 * @function bindAll
 * helper for binding prototype methods to class context
 *
 * for example, instead of doing this in a class `constructor`:
 * `this.handleClick = this.handleClick.bind(this)`
 *
 * you can use `bindAll`:
 * ```
 * bindAll(this,
 * 	handleClick,
 * 	someOtherMethod,
 * 	anotherMethod);
 * ```
 */
export const bindAll = (context, ...names) => {
	names.forEach(name => context[name] = context[name].bind(context));
};
