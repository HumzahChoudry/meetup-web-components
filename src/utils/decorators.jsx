import React from 'react';
import { IntlProvider } from 'react-intl';

export const decorateWithLocale = story => {
	const locale = 'en-US';

	/*
	 * -- Inline SVG icon sprite --
	 *
	 * raw SVG sprite from `swarm-icons`
	 */
	const iconSpriteStyle = { display: 'none' };
	const iconSprite = require('raw-loader!swarm-icons/dist/sprite/sprite.inc');
	const styles = {
		display:'flex',
		alignItems: 'center',
		justifyContent: 'center',
		height: '100%',
		width: '100%'
	};

	return (
		<IntlProvider locale={locale}>
			<div style={styles}>
				<div style={iconSpriteStyle} dangerouslySetInnerHTML={{__html: iconSprite}} />
				{story()}
			</div>
		</IntlProvider>
	);
};
