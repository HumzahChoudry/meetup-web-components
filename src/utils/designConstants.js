export const SIZE_SCALE_RATIO = 1.125;

export const BREAKPOINTS = {
	s: '440px',
	m: '640px',
	l: '840px',
	xl: '1024px'
};

export const MEDIA_SIZES = {
	xs: 16,
	s: 24,
	m: 36,
	l: 48,
	xl: 72,
	xxl: 120
};

export function changeSizeWithViewport(size, breakpoint) {
	const breakpointSizes = {
		s: MEDIA_SIZES[size],
		m: Math.floor(MEDIA_SIZES[size] * SIZE_SCALE_RATIO),
		l: Math.floor(MEDIA_SIZES[size] * Math.pow(SIZE_SCALE_RATIO, 2)),
		xl: Math.floor(MEDIA_SIZES[size] * Math.pow(SIZE_SCALE_RATIO, 2))
	};
	return breakpointSizes[breakpoint];
}

