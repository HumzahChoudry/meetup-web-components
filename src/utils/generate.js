const fs = require('fs');
const path = require('path');
const inquirer = require('inquirer');

require('babel-register');  // process all further imports through babel

function validateName(name) {
	return /^\b([A-Z][a-z]*)/.test(name) ? true : 'Name must be TitleCase';
}

function promptCreate() {
	const questions = [
		{
			type: 'input',
			name: 'name',
			message: (answers) => 'Name of your component:',
			validate: validateName
		}, {
			type: 'confirm',
			name: 'isIntl',
			message: (answers) => 'Add i18n wrappers to tests and story?',
			default: false
		}
	];

	return inquirer.prompt(questions);
}

// Begin prompts
promptCreate()
	.then(answers => ({
		answers,
		createPath: path.resolve(__dirname, '../../src/')
	}))
	.then((config) => {
		'use strict';
		const createPath = config.createPath;
		const name = config.answers.name;
		const isIntl = config.answers.isIntl;
		const nameLcase = name.replace(/^[A-Z]/, (firstLetter) => firstLetter.toLowerCase());
		let filepath;

		filepath = path.resolve(createPath, `${name}.jsx`);
		console.log('Writing component', filepath);
		fs.writeFileSync(
			filepath,
			require('./templates/component.jsx').default(name)
		);

		filepath = path.resolve(createPath, `${nameLcase}.test.jsx`);
		console.log('Writing component test', filepath);
		fs.writeFileSync(
			filepath,
			require('./templates/test.jsx').default(name, isIntl)
		);

		filepath = path.resolve(createPath, `${nameLcase}.story.jsx`);
		console.log('Writing component story', filepath);
		fs.writeFileSync(
			filepath,
			require('./templates/story.jsx').default(name, isIntl)
		);
	})
	.then(
		() => {
			console.log('Done');
		},
		(err) => {
			console.log(err.stack);
		}
	);
