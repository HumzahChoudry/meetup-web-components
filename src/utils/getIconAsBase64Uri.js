export const getIconAsBase64Uri = name => {
	const iconPath = `~swarm-icons/dist/optimized/${name}.svg`;
	const base64Content = `base64-loader!${iconPath}`;

	console.log(iconPath);

	return `data:image/svg+xml;charset=utf-8;base64,${base64Content}`;
};
