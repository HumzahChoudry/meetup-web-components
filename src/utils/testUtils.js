import React from 'react';
import { shallow } from 'enzyme';
// import TestUtils from 'react-dom/test-utils';

export const variantTest = (FoundationComponent, className, variants) => {
	variants.forEach(variant => {
		const props = {
			[variant]: true,
		};

		const component = shallow(<FoundationComponent {...props} />);
		expect(component.find(`.${className}--${variant}`).exists()).toBe(true);
	});
};

export const hasChildByClassName = (el, className) => {
	const hasClass = !!el.getElementsByClassName(className).length;
	expect(hasClass).toBe(true);
};

export const hasAttribute = (el, attr, value) => {
	const elAttrValue = el.getAttribute(attr);
	expect(elAttrValue).not.toBeNull();
	if (value !== undefined) {
		expect(` ${elAttrValue} `.indexOf(` ${value} `)).not.toBe(-1);
	}
};

export const hasRoleAttribute = (el, roleName) => {
	hasAttribute(el, 'role', roleName);
};

export const componentHasProperty = (component, prop, value) => {
	expect(component && component.props && component.props[prop] === value).toBe(
		true
	);
};
