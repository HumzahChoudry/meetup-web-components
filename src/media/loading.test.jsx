import React from 'react';
import { variantTest } from '../utils/testUtils';
import { shallow } from 'enzyme';

import Loading, {
	LOADING_CLASS,
	LOADING_SHAPE_CLASS,
	LOADING_SHAPE_PATH_CLASS
} from './Loading';

const CUSTOM_COLOR = '#000';
const CUSTOM_SCRIM_COLOR = '#FFF';
const CUSTOM_SIZE = '42px';

describe('Loading', () => {
	const loadingComponent = shallow(
		<Loading
			fullCover
			color={CUSTOM_COLOR}
			scrimColor={CUSTOM_SCRIM_COLOR}
			size={CUSTOM_SIZE}
		/>
	);
	it('exists', () => {
		expect(loadingComponent).toMatchSnapshot();
	});

	it('applies variant classes for each variant prop', () => {
		const variants = [
			'fullCover',
			'partialCover'
		];

		variantTest(Loading, LOADING_CLASS, variants);
	});

	it('colors the loader', () => {
		const shapePath = loadingComponent.find(`.${LOADING_SHAPE_PATH_CLASS}`);
		expect(shapePath.prop('stroke')).toBe(CUSTOM_COLOR);
	});

	it('colors the scrim', () => {
		expect(loadingComponent.prop('style').backgroundColor).toBe(CUSTOM_SCRIM_COLOR);
	});

	it('resizes when a custom size is passed', () => {
		const shape = loadingComponent.find(`.${LOADING_SHAPE_CLASS}`);
		expect(shape.prop('style').width).toBe(CUSTOM_SIZE);
	});
});
