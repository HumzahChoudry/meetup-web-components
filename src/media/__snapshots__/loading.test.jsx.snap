// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading exists 1`] = `
<div
  className="loading loading--cover loading--fullCover"
  style={
    Object {
      "backgroundColor": "#FFF",
    }
  }
>
  <svg
    aria-labelledby="title"
    className="loadingShape"
    role="progressbar"
    style={
      Object {
        "width": "42px",
      }
    }
    viewBox="0 0 50 50"
  >
    <title
      className="visibility--a11yHide"
      id="title"
    >
      Loading
    </title>
    <circle
      className="loadingShape-path"
      cx="25"
      cy="25"
      fill="none"
      r="20"
      stroke="#000"
      stroke-miterlimit="10"
      stroke-width="3"
    />
  </svg>
</div>
`;
